### PR TITLE
Error on private items with stability attributes

### LINKED ISSUE #83380: Rustc should warn if you have a stability attribute on a non-public item
_Originally posted by @ijackson in https://github.com/rust-lang/rust/issues/83372#issuecomment-804135923_

### LINKED ISSUE #83396: meta: Lint against items that have stability attributes but are private
Adding this would likely have prevented https://github.com/rust-lang/rust/pull/83372#issuecomment-804135923.

### PR DESCRIPTION
**Pull request by camelid**
Fixes https://github.com/rust-lang/rust/issues/83380.

It doesn't make sense for a private item to have a stability attribute,
and it is likely a mistake. Report an error when we see a case of this.

Inspired by https://github.com/rust-lang/rust/pull/83372#issuecomment-804135923.

Also tweaked wording of "missing stability attribute" error.

### CODE REVIEW

**Comment by crlf0710** [comment]
@camelid Ping from triage, what's the current status of this? Seems CI is still red.

**Comment by camelid** [comment]
> @camelid Ping from triage, what's the current status of this? Seems CI is still red.

Yeah, I'm not really sure why the code isn't working, and I haven't gotten around to asking yet.

**Comment by m-ou-se** [comment]
> should we do this at all

I'm very much in favour of lints like this for the attributes we use in `std`. It's easy to make mistakes with these attributes, and the more easily an error is caught the better.

I haven't looked in detail through the errors CI is giving, but I wonder how many of these are actual errors with these attributes in `std` that we only discover now.

Assigning it back for compiler review:

r? @matthewjasper

File: compiler/rustc_passes/src/stability.rs [hunk 1]
```diff
@@ -559,6 +559,24 @@ impl<'tcx> MissingStabilityAnnotations<'tcx> {
             }
         }
     }
+
+    fn check_private_stability(&self, hir_id: HirId, span: Span) {
+        let stab = self.tcx.stability().local_stability(hir_id);
+        let is_error = stab.is_some() && !self.access_levels.is_reachable(hir_id);
```
Comment (jyn514) [inline compiler/rustc_passes/src/stability.rs:565]: Note this is currently buggy: https://github.com/rust-lang/rust/issues/64762

File: compiler/rustc_passes/src/stability.rs [hunk 2]
```diff
@@ -559,6 +559,24 @@ impl<'tcx> MissingStabilityAnnotations<'tcx> {
             }
         }
     }
+
+    fn check_private_stability(&self, hir_id: HirId, span: Span) {
+        let stab = self.tcx.stability().local_stability(hir_id);
```
Comment (jyn514) [inline compiler/rustc_passes/src/stability.rs:564]: I don't think this is the right check; try using [`lookup_stability`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.lookup_stability) instead.
```suggestion
        let stab = self.tcx.lookup_stability(hir_id);
```

**Comment by crlf0710** [comment]
@camelid Ping from triage, CI still red, and there's merge conflicts. Would you mind addressing the comments above? Thanks!

If you need help from team members, maybe Zulip is a more effective approach to talk with them. Thanks!

**Comment by camelid** [comment]
Starting with a rebase...

**Comment by crlf0710** [comment]
@camelid Ping from triage, CI is still red here.

**Comment by JohnCSimon** [comment]
@camelid Ping  agiain from triage, CI is still failing

**Comment by camelid** [comment]
I will try to get to this soon.

**Comment by camelid** [comment]
Rebased.

**Comment by camelid** [comment]
Hmm, still getting thousands of stability errors when compiling `core`...

**Comment by camelid** [comment]
Ok, it seems like the issue is that definitions are inheriting stability from their parent. E.g., in this code:

```rust
#![stable(feature = "rust1", since = "1.0.0")]

#[stable(feature = "rust1", since = "1.0.0")]
pub struct MyStruct {
    my_field: u32,
}
```

`MyStruct.my_field` is inheriting `MyStruct`'s stability attribute, which causes this error:

```
error: field is private but has a stability attribute
 --> foo.rs:8:5
  |
8 |     my_field: u32,
  |     ^^^^^^^^^^^^^
  |
  = help: if it is meant to be private, remove the stability attribute
  = help: or, if it is meant to be public, make it public
```

Removing both the crate-level stability attribute and the struct-level one causes the field error to disappear (although then it rightly errors because the crate and struct are public but don't have stability attributes). It seems like this might be a pre-existing bug in how stabilities are computed.

**Comment by camelid** [comment]
Ok, I fixed part of it: it's down from 3608 to 3300 errors 😅 

I think the bugs might have been introduced in #71481.

**Comment by camelid** [comment]
It seems that some of the errors were due to bugs in the stability-checking code, but many of them (probably the vast majority) were because some items were private but should still inherit stability from their parent. I'm pretty sure that the stability of a private item is irrelevant because it can't be accessed outside the crate, but I'd rather not mess with the existing stability code too much since it's a bit hard to understand what's going on. So, I think a better approach, at least for now, is to just look at the item's actual stability *attributes* (not what it's inherited from its parent). This approach seems promising: I'm down to 53 errors in `core`! Next, I need to see if they are false positives or actually correct.

**Comment by camelid** [comment]
A lot of the errors seem to be correct! There are a few false positives; so far, I've seen a couple cases where `derive`d impls trigger the error. I'm not sure why it only happens in a few cases.

**Comment by camelid** [comment]
>  I'm pretty sure that the stability of a private item is irrelevant because it can't be accessed outside the crate, but I'd rather not mess with the existing stability code too much since it's a bit hard to understand what's going on. So, I think a better approach, at least for now, is to just look at the item's actual stability _attributes_ (not what it's inherited from its parent).

Hmm, I wonder if a better approach would be to not inherit stability for private items...

**Comment by camelid** [comment]
> A lot of the errors seem to be correct! There are a few false positives; so far, I've seen a couple cases where `derive`d impls trigger the error. I'm not sure why it only happens in a few cases.

Actually those weren't false positives: the `derive`s were on a private struct with stability attributes.

**Comment by camelid** [comment]
Unfortunately, now that I've removed the `core` stability attributes that were on private items, I'm getting thousands of errors when compiling stage0 `core`, this time that the items are missing stability attributes. It seems that the stability-checker thinks that private items need attributes?

**Comment by jyn514** [comment]
@camelid can you post one of the errors?

**Comment by camelid** [comment]
Sure, here are some samples:

<details>
<summary>Sample 1 (errors from the start of the output)</summary>

```
error: trait has missing stability attribute
  --> library/core/src/ptr/metadata.rs:51:1
   |
51 | / pub trait Pointee {
52 | |     /// The type for metadata in pointers and references to `Self`.
53 | |     #[lang = "metadata_type"]
54 | |     // NOTE: Keep trait bounds in `static_assert_expected_bounds_for_metadata`
...  |
57 | |     type Metadata: Copy + Send + Sync + Ord + Hash + Unpin;
58 | | }
   | |_^

error: function has missing stability attribute
  --> library/core/src/ptr/metadata.rs:91:1
   |
91 | / pub const fn metadata<T: ?Sized>(ptr: *const T) -> <T as Pointee>::Metadata {
92 | |     // SAFETY: Accessing the value from the `PtrRepr` union is safe since *const T
93 | |     // and PtrComponents<T> have the same memory layouts. Only std can make this
94 | |     // guarantee.
95 | |     unsafe { PtrRepr { const_ptr: ptr }.components.metadata }
96 | | }
   | |_^

error: struct has missing stability attribute
   --> library/core/src/ptr/metadata.rs:176:1
    |
176 | / pub struct DynMetadata<Dyn: ?Sized> {
177 | |     vtable_ptr: &'static VTable,
178 | |     phantom: crate::marker::PhantomData<Dyn>,
179 | | }
    | |_^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:213:1
    |
213 | unsafe impl<Dyn: ?Sized> Send for DynMetadata<Dyn> {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:214:1
    |
214 | unsafe impl<Dyn: ?Sized> Sync for DynMetadata<Dyn> {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:216:1
    |
216 | / impl<Dyn: ?Sized> fmt::Debug for DynMetadata<Dyn> {
217 | |     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
218 | |         f.debug_tuple("DynMetadata").field(&(self.vtable_ptr as *const VTable)).finish()
219 | |     }
220 | | }
    | |_^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:224:1
    |
224 | impl<Dyn: ?Sized> Unpin for DynMetadata<Dyn> {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:226:1
    |
226 | impl<Dyn: ?Sized> Copy for DynMetadata<Dyn> {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:228:1
    |
228 | / impl<Dyn: ?Sized> Clone for DynMetadata<Dyn> {
229 | |     #[inline]
230 | |     fn clone(&self) -> Self {
231 | |         *self
232 | |     }
233 | | }
    | |_^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:235:1
    |
235 | impl<Dyn: ?Sized> Eq for DynMetadata<Dyn> {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:237:1
    |
237 | / impl<Dyn: ?Sized> PartialEq for DynMetadata<Dyn> {
238 | |     #[inline]
239 | |     fn eq(&self, other: &Self) -> bool {
240 | |         crate::ptr::eq::<VTable>(self.vtable_ptr, other.vtable_ptr)
241 | |     }
242 | | }
    | |_^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:244:1
    |
244 | / impl<Dyn: ?Sized> Ord for DynMetadata<Dyn> {
245 | |     #[inline]
246 | |     fn cmp(&self, other: &Self) -> crate::cmp::Ordering {
247 | |         (self.vtable_ptr as *const VTable).cmp(&(other.vtable_ptr as *const VTable))
248 | |     }
249 | | }
    | |_^

error: implementation has missing stability attribute
   --> library/core/src/ptr/metadata.rs:251:1
    |
251 | / impl<Dyn: ?Sized> PartialOrd for DynMetadata<Dyn> {
252 | |     #[inline]
253 | |     fn partial_cmp(&self, other: &Self) -> Option<crate::cmp::Ordering> {
254 | |         Some(self.cmp(other))
255 | |     }
256 | | }
    | |_^
```
</details>

<details>
<summary>Sample 2 (errors from the end of the output)</summary>

```
error: implementation has missing stability attribute
   --> library/core/src/../../stdarch/crates/core_arch/src/macros.rs:88:18
    |
82  | /  macro_rules! types {
83  | |      ($(
84  | |          $(#[$doc:meta])*
85  | |          pub struct $name:ident($($fields:tt)*);
...   |
88  | |          #[derive(Copy, Clone, Debug)]
    | |                   ^^^^ in this macro invocation (#2)
...   |
93  | |      )*)
94  | |  }
    | |__- in this expansion of `types!` (#1)
    | 
   ::: library/core/src/../../stdarch/crates/core_arch/src/x86/mod.rs:8:1
    |
8   |  / types! {
9   |  |     /// 128-bit wide integer vector type, x86-specific
10  |  |     ///
11  |  |     /// This type is the same as the `__m128i` type defined by Intel,
...    |
329 |  |     );
330 |  | }
    |  |_- in this macro invocation (#1)
    | 
   ::: library/core/src/marker.rs:393:1
    |
393 | /  pub macro Copy($item:item) {
394 | |      /* compiler built-in */
395 | |  }
    | |__- in this expansion of `#[derive(Copy)]` (#2)

error: implementation has missing stability attribute
   --> library/core/src/../../stdarch/crates/core_arch/src/macros.rs:88:24
    |
82  | /   macro_rules! types {
83  | |       ($(
84  | |           $(#[$doc:meta])*
85  | |           pub struct $name:ident($($fields:tt)*);
...   |
88  | |           #[derive(Copy, Clone, Debug)]
    | |                          ^^^^^ in this macro invocation (#2)
...   |
93  | |       )*)
94  | |   }
    | |___- in this expansion of `types!` (#1)
    | 
   ::: library/core/src/../../stdarch/crates/core_arch/src/x86/mod.rs:8:1
    |
8   |   / types! {
9   |   |     /// 128-bit wide integer vector type, x86-specific
10  |   |     ///
11  |   |     /// This type is the same as the `__m128i` type defined by Intel,
...     |
329 |   |     );
330 |   | }
    |   |_- in this macro invocation (#1)
    | 
   ::: library/core/src/clone.rs:139:1
    |
139 | /   pub macro Clone($item:item) {
140 | |       /* compiler built-in */
141 | |   }
    | |___- in this expansion of `#[derive(Clone)]` (#2)

error: implementation has missing stability attribute
   --> library/core/src/../../stdarch/crates/core_arch/src/macros.rs:88:31
    |
82  | /  macro_rules! types {
83  | |      ($(
84  | |          $(#[$doc:meta])*
85  | |          pub struct $name:ident($($fields:tt)*);
...   |
88  | |          #[derive(Copy, Clone, Debug)]
    | |                                ^^^^^ in this macro invocation (#2)
...   |
93  | |      )*)
94  | |  }
    | |__- in this expansion of `types!` (#1)
    | 
   ::: library/core/src/../../stdarch/crates/core_arch/src/x86/mod.rs:8:1
    |
8   |  / types! {
9   |  |     /// 128-bit wide integer vector type, x86-specific
10  |  |     ///
11  |  |     /// This type is the same as the `__m128i` type defined by Intel,
...    |
329 |  |     );
330 |  | }
    |  |_- in this macro invocation (#1)
    | 
   ::: library/core/src/fmt/mod.rs:618:5
    |
618 | /      pub macro Debug($item:item) {
619 | |          /* compiler built-in */
620 | |      }
    | |______- in this expansion of `#[derive(Debug)]` (#2)

error: implementation has missing stability attribute
   --> library/core/src/../../stdarch/crates/core_arch/src/macros.rs:88:18
    |
82  | /  macro_rules! types {
83  | |      ($(
84  | |          $(#[$doc:meta])*
85  | |          pub struct $name:ident($($fields:tt)*);
...   |
88  | |          #[derive(Copy, Clone, Debug)]
    | |                   ^^^^ in this macro invocation (#2)
...   |
93  | |      )*)
94  | |  }
    | |__- in this expansion of `types!` (#1)
    | 
   ::: library/core/src/../../stdarch/crates/core_arch/src/x86/mod.rs:8:1
    |
8   |  / types! {
9   |  |     /// 128-bit wide integer vector type, x86-specific
10  |  |     ///
11  |  |     /// This type is the same as the `__m128i` type defined by Intel,
...    |
329 |  |     );
330 |  | }
    |  |_- in this macro invocation (#1)
    | 
   ::: library/core/src/marker.rs:393:1
    |
393 | /  pub macro Copy($item:item) {
394 | |      /* compiler built-in */
395 | |  }
    | |__- in this expansion of `#[derive(Copy)]` (#2)

error: implementation has missing stability attribute
   --> library/core/src/../../stdarch/crates/core_arch/src/macros.rs:88:24
    |
82  | /   macro_rules! types {
83  | |       ($(
84  | |           $(#[$doc:meta])*
85  | |           pub struct $name:ident($($fields:tt)*);
...   |
88  | |           #[derive(Copy, Clone, Debug)]
    | |                          ^^^^^ in this macro invocation (#2)
...   |
93  | |       )*)
94  | |   }
    | |___- in this expansion of `types!` (#1)
    | 
   ::: library/core/src/../../stdarch/crates/core_arch/src/x86/mod.rs:8:1
    |
8   |   / types! {
9   |   |     /// 128-bit wide integer vector type, x86-specific
10  |   |     ///
11  |   |     /// This type is the same as the `__m128i` type defined by Intel,
...     |
329 |   |     );
330 |   | }
    |   |_- in this macro invocation (#1)
    | 
   ::: library/core/src/clone.rs:139:1
    |
139 | /   pub macro Clone($item:item) {
140 | |       /* compiler built-in */
141 | |   }
    | |___- in this expansion of `#[derive(Clone)]` (#2)

error: implementation has missing stability attribute
   --> library/core/src/../../stdarch/crates/core_arch/src/macros.rs:88:31
    |
82  | /  macro_rules! types {
83  | |      ($(
84  | |          $(#[$doc:meta])*
85  | |          pub struct $name:ident($($fields:tt)*);
...   |
88  | |          #[derive(Copy, Clone, Debug)]
    | |                                ^^^^^ in this macro invocation (#2)
...   |
93  | |      )*)
94  | |  }
    | |__- in this expansion of `types!` (#1)
    | 
   ::: library/core/src/../../stdarch/crates/core_arch/src/x86/mod.rs:8:1
    |
8   |  / types! {
9   |  |     /// 128-bit wide integer vector type, x86-specific
10  |  |     ///
11  |  |     /// This type is the same as the `__m128i` type defined by Intel,
...    |
329 |  |     );
330 |  | }
    |  |_- in this macro invocation (#1)
    | 
   ::: library/core/src/fmt/mod.rs:618:5
    |
618 | /      pub macro Debug($item:item) {
619 | |          /* compiler built-in */
620 | |      }
    | |______- in this expansion of `#[derive(Debug)]` (#2)

error: associated type has missing stability attribute
  --> library/core/src/ptr/metadata.rs:57:5
   |
57 |     type Metadata: Copy + Send + Sync + Ord + Hash + Unpin;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: associated function has missing stability attribute
   --> library/core/src/ptr/metadata.rs:194:5
    |
194 | /     pub fn size_of(self) -> usize {
195 | |         self.vtable_ptr.size_of
196 | |     }
    | |_____^

error: associated function has missing stability attribute
   --> library/core/src/ptr/metadata.rs:200:5
    |
200 | /     pub fn align_of(self) -> usize {
201 | |         self.vtable_ptr.align_of
202 | |     }
    | |_____^

error: associated function has missing stability attribute
   --> library/core/src/ptr/metadata.rs:206:5
    |
206 | /     pub fn layout(self) -> crate::alloc::Layout {
207 | |         // SAFETY: the compiler emitted this vtable for a concrete Rust type which
208 | |         // is known to have a valid layout. Same rationale as in `Layout::for_value`.
209 | |         unsafe { crate::alloc::Layout::from_size_align_unchecked(self.size_of(), self.align_of()) }
210 | |     }
    | |_____^

error: aborting due to 3847 previous errors

error: could not compile `core`
```
</details>

**Comment by jyn514** [comment]
You said:
> I'm getting thousands of errors when compiling stage0 core, this time that the items are missing stability attributes. It seems that the stability-checker thinks that private items need attributes?

But all the errors are on public items. Maybe the problem is that some of the items are not actually private, because they're re-exported somewhere? If they're not private, then it seems worth making a cleanup PR to remove `pub`.

**Comment by camelid** [comment]
I'm not sure why this change doesn't work correctly, and I probably won't be able to look into it soon, so closing for now. I may try again in the future if someone else doesn't do it by then.

**Pull request closed without merging**